### ISSUE TITLE
Expose FileVersionProvider.GetHashForFile (with caching)

### ISSUE THREAD

**Issue by q42jaap** [aspnet/Mvc#7735]
#### Is this a Bug or Feature request?:
Feature request

#### Description of the problem:
We're using a multi zone azure setup where we deploy first to 1 datacenter and then to the next. All traffic is handled by 1 azure traffic manager/loadbalancer (I'm leaving out how we deploy to a staging slot first, since I'm not sure how it is exactly done, and I doubt whether this is relevant).
The site uses a CDN for it's static files. So the HTML links to some azureedge.net instance for it's javascript/css.

We use the `FileVersionProvider` in the `_Layout.cshtml` to generate script tags and stylesheet links (https://github.com/aspnet/Mvc/blob/2.1.0-preview2-final/src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/FileVersionProvider.cs).
So far so good, we get nice cachable URLs which directly expire when the content changes.

We've configured the static file middleware to add cache headers like so:
```
app.UseStaticFiles(new StaticFileOptions
{
	OnPrepareResponse = context =>
	{
		var responseHeaders = context.Context.Response.GetTypedHeaders();
		responseHeaders.CacheControl = new CacheControlHeaderValue
		{
			MaxAge = TimeSpan.FromDays(30),
			Public = true
		};
	}
});
```
This way the `v` querystring param works very well, when deploying a small change which didn't involve javascript/css changes, the browser cache can just use whatever it has stored.

#### What happens when we do change javascript or css.
While an active deployment is going on, we first deploy to the 1st zone (we swap slots) and some time later we deploy to the 2nd zone (could be just a couple of minutes, or a little bit longer if we need to run checks first).
After all is well we have to purge the cache.

The CDN caches requests regardless of the `v` version key (see Motivation), so when it encounters a call script.js?v=a3fe.... it will do the lookup based on script.js. Because we purged the CDN after both zones are deployed we'll get the new version. However, if somebody visited the site while the deploy was in progress, the cache propagates to the end user's browser. 

While the deploy is running, the static file handlers will just serve the request regardless of what the `v` query param is. So it might be the old version. And this is cached by the CDN, so it gets polluted.
We eventually clear the cache but then the damage is already done, old content is cached by the browser under the new version hash. Purging the CDN fixes this for future visitors, but them who happened to visit during that deploy have a stale version for 30 days.

*Took us some time to figure this out*

#### The fix
We wanted to adjust our caching strategy:
```
app.UseStaticFiles(new StaticFileOptions
{
	OnPrepareResponse = context =>
	{
		var requestVersionHash = context.Context.Request.Query["v"].FirstOrDefault();
		var cacheDuration = TimeSpan.FromDays(30);
		if (!string.IsNullOrEmpty(requestVersionHash))
		{
			var cdnUrlService = context.Context.RequestServices.GetService<CdnUrlService>();
			var currentVersionHash = cdnUrlService.GetVersionHash(context.File);
			if (currentVersionHash != requestVersionHash)
			{
				// we don't disable caching entirely, but set it to a short period so stale content won't end up in the cache for a long time. 
				cacheDuration = TimeSpan.FromMinutes(1); 
			}
		}

		var responseHeaders = context.Context.Response.GetTypedHeaders();
		responseHeaders.CacheControl = new CacheControlHeaderValue
		{
			MaxAge = cacheDuration,
			Public = true
		};
	}
});
```

The CdnUrlService is something we cooked up ourselves, and it uses the FileVersionProvider (_Layout.cshtml calls the CdnUrlService to add `v` to the paths).

#### The problem and (finally) the feature request
Our method that gets the version hash can't be implemented by calling some method of FileVersionProvider.
```
cdnUrlService.GetVersionHash(context.File);
```

It would need to call `private static string GetHashForFile(IFileInfo fileInfo)` but it's private. Further more, that method is private for a reason, because it doesn't cache, the method `AddFileVersionToPath` also caches the value.

My request would be to better support the statics-via-cdn-with-multiple-zones scenario.
One way which would work for us is to expose a new instance method on the `FileVersionProvider` that takes an IFileInfo and handles the caching, `AddFileVersionToPath` should then use that method.
We could then use that method to verify the versionHash in the static file handlers to adjust cache duration.

Then there is the issue of FileVersionProvider to be an `Internal` namespace, we are aware that the class is not intended to be used by users, but the same problems still persist. There is no way (without copying the code in FileVersionProvider, yay open source!) to check the `v` param inside the static file handler's `OnPrepareResponse`.

#### Motivation
The CDN we're using is Akamai, which doesn't support rules which could fix this, by including the `v` query string in the cache key. Only the premium tier supports this taking `v` into account. The `FileVersionProvider` thus only works well with a premium tiered CDN, or with flushing CDN cache at the right times. However, when using multiple zones, there is no right time to flush the cache, since it will get polluted and propagated to browsers.
Our proposed fix will increase traffic a bit during a deploy which is just fine. Having the internals of the FileVersionProvider exposed into a usable API could be beneficial in other scenarios as well (other users might want to return 404 when version hashes mismatch).

#### Remarks
There may be other ways to coordinate the deployment, I'm not sure if you wouldn't run into the same issues. Fact still is that FileVersionProvider doesn't work well in CDN scenario's where query params aren't used/configured into the cache key.
We might just be better off with the premium CDN, which we'll be discussing. Switching to premium just to fix this, does seem like overkill though.

**Comment by q42jaap** [comment]
*Update:*
Even with the Verizon CDN, we experience this problem. The new html references the new versionHash, but could still be handled by an older instance of the app, or the 2nd datacenter. We've come up with our own implementation of a `FileVersionProvider`-like class.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us with this issue, @q42jaap.
I do agree that the proper fix here would be on the CDN side, to respect the `version` info.

@rynowak, @pranavkm do you think providing extensibility for the `FileVersionProvider` is a feasible option for us? Or maybe you'd have a better suggestion here?

**Comment by q42jaap** [comment]
@mkArtakMSFT Please not that with the verizon CDN, premium or standard, this problem still exists. The way we use the CDN in front of the traffic manager to handle the traffic for our static files causes these problems.
The Verizon CDN can be configured to respect the version hash in the query string, no need to switch to premium for that.

**Comment by pranavkm** [comment]
The [recommendation](https://stackoverflow.com/a/9692773) is to embed file versions in their name rather than in a query string. IMO, this seems like a much nicer solution to trying to mess around with the cache headers. My suggestion involves

1) a publish target that publishes files with a version e.g. `site.{hash}.css` 
b) writes the mapping of original file names to a file

The tag helpers would look up paths in these manifests to produce the link, falling back to versions in query string if we want to support cache busting for dev environments. This also allows for doing side-by-side versioning of files in case there's ever a need to support multiple file versions.

**Comment by q42jaap** [comment]
@pranavkm I still have to add that such a manifest and the infrastructure around reading out these values (probably want to cache them like `FileVersionProvider`) needs to be more open than currently is the case. 

Furthermore, a CDN with a trafficManager behind it will not know which server to address to actually get that unique filenam, which would cause 404's on css/js during a rolling deploy. In that case we would be better off with the current setup where a query string parameter is used. The side-by-side versioning is really difficult to guarantee in a, for example, Docker-type deployment scenario.

In the query string scenario static file middleware (`OnPrepareResponse`) should be able to plug in to the internals of `FileVersionProvider`.

I've created a gist of our current implementation for reference: https://gist.github.com/q42jaap/377d1a2b060c017b244a6a5e250585c4

**Comment by pranavkm** [comment]
> In that case we would be better off with the current setup where a query string parameter is used.

You might be serving incompatible js or css, that doesn't sound optimal. But I'm not going to question your deployment experience. 

We already have an issue - https://github.com/aspnet/Mvc/issues/6371- to make the `FileVersionProvider` interface a contract and that's marked as up-for-grabs. It seems like if you had access to the contract, it should unblock you. Would you like to consider contributing to the other issue?

**Comment by q42jaap** [comment]
Yes the incompatible js or css is less than ideal, unless the Azure traffic manager is going to support CDN capabilities internally, we're stuck in this situation. Taking 1 zone out of the traffic manager also isn't optimal.

I'd be willing to contribute to #6371 how should we proceed?

**Comment by pranavkm** [comment]
Closing since we're tracking making the API public as part of https://github.com/aspnet/Mvc/issues/6371.
